### PR TITLE
feat: End to end refinance

### PR DESCRIPTION
**Pull request by JGcarv**

### CODE REVIEW

File: contracts/test/RefinanceTest.t.sol [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+        // Refinance parameters
+
+        uint256[3] memory refinanceTermDetails = [
+            uint256(15 days),  // 15 day grace period
+            uint256(45 days),  // 45 day payment interval
+            uint256(3)
+        ];
+
+        // 5 BTC @ ~$58k = $290k = 29% collateralized, interest only
```
Comment (lucas-manuel) [inline contracts/test/RefinanceTest.t.sol:230]: Need to update comment for 2BTC

File: contracts/test/RefinanceTest.t.sol [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+        borrower.loan_proposeNewTerms(address(loanV2), address(refinancer), data);
+
+        // Pool Delegate funds loan again for increasing amount
+        pool.fundLoan(address(loanV2), address(debtLockerFactory), principalIncrease);
+
+        // Pool Delegate(address this) accepts new terms on Debt Locker
+        DebtLocker debtLocker = DebtLocker(loanV2.lender());
+
+        debtLocker.acceptNewTerms(address(refinancer), data, principalIncrease);
```
Comment (lucas-manuel) [inline contracts/test/RefinanceTest.t.sol:258]: We should assert that this fails before the fundLoan is called with a try/catch

File: contracts/test/RefinanceTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        assertEq(loanV2.paymentsRemaining(), refinanceTermDetails[2]);
+
+        assertEq(loanV2.principalRequested(), requests[1] + principalIncrease);
+        assertEq(loanV2.principal(),          requests[1] + principalIncrease);
+        assertEq(loanV2.collateralRequired(), refinanceRequests[0]);
+
+        assertEq(loanV2.interestRate(),        refinanceRates[0]);
+        assertEq(loanV2.lateFeeRate(),         refinanceRates[2]);
+        assertEq(loanV2.lateInterestPremium(), refinanceRates[3]);
+
```
Comment (lucas-manuel) [inline contracts/test/RefinanceTest.t.sol:271]: Lets do another payment and claim after the refinance and explictly assert the values to show a change in the amount

Comment (lucas-manuel) [reply contracts/test/RefinanceTest.t.sol:271]: Lets also set endingPrincipal to be lower than 2m to demonstrate that we can move from interest-only to amortized using a refiannce

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Looks good! Some suggestions

File: contracts/test/RefinanceTest.t.sol [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        // 5 BTC @ ~$58k = $290k = 29% collateralized, interest only
+        uint256[3] memory requests = [uint256(5 * BTC), uint256(1_000_000 * USD), uint256(1_000_000 * USD)];  
+
+        uint256[4] memory rates = [uint256(0.12e18), uint256(0), uint256(0.05e18), uint256(0.6e18)]; 
+
+        bytes memory arguments = loanInitializer.encodeArguments(address(borrower), assets, termDetails, requests, rates);
+
+        loanV2 = IMapleLoan(borrower.mapleProxyFactory_createInstance(address(loanFactory), arguments));
+
+        /*****************/
```
[the remark this answers is not in the document]
Comment (JGcarv) [reply contracts/test/RefinanceTest.t.sol:145]: I don't understand what do you mean here

[the remark this answers is not in the document]
Comment (lucas-manuel) [reply contracts/test/RefinanceTest.t.sol:145]: Like move
```js
/*****************/
/*** Fund Loan ***/
/*****************/
```
into the scoping block so its aligned with the drawdown label. The way it looks currently looks like the whole scoping block is for funding the loan

File: contracts/test/RefinanceTest.t.sol [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+        // Refinance parameters
+
+        uint256[3] memory refinanceTermDetails = [
+            uint256(15 days),  // 15 day grace period
+            uint256(45 days),  // 45 day payment interval
+            uint256(3)
+        ];
+
+        // 2 BTC @ ~$58k = $290k = 29% collateralized, partially amortized
```
Comment (lucas-manuel) [inline contracts/test/RefinanceTest.t.sol:206]: `2 BTC @ ~$58k = $116k = 11.6% collateralized, partially amortized`

**Pull request merged**